src/render.c

```diff
@@ -0,0 +1,226 @@
+    {.y = 13, .x = 39},  // SHAPE_I
+    {.y = 0, .x = 47},   // SHAPE_L
+    {.y = 0, .x = 67},   // SHAPE_N
+    {.y = 4, .x = 47},   // SHAPE_P
+    {.y = 9, .x = 62},   // SHAPE_T
+    {.y = 9, .x = 48},   // SHAPE_U
+    {.y = 4, .x = 54},   // SHAPE_V
+    {.y = 9, .x = 56},   // SHAPE_W
+    {.y = 4, .x = 68},   // SHAPE_X
+    {.y = 0, .x = 57},   // SHAPE_Y
+    {.y = 4, .x = 61},   // SHAPE_Z
+};
+
+const shape_t tile_relation[4][6] = {
+    {SHAPE_E, SHAPE_E, SHAPE_L, SHAPE_Y, SHAPE_Y, SHAPE_N},
+    {SHAPE_J, SHAPE_S, SHAPE_P, SHAPE_V, SHAPE_Z, SHAPE_X},
+    {SHAPE_H, SHAPE_Q, SHAPE_U, SHAPE_W, SHAPE_T, SHAPE_F},
+    {SHAPE_I, SHAPE_I, SHAPE_A, SHAPE_R, SHAPE_D, SHAPE_M}
+};
+
+int init_all_colors() {
+    start_color();
+    init_pair(RED_PAIR, COLOR_RED, COLOR_BLACK);
+    init_pair(GREEN_PAIR, COLOR_GREEN, COLOR_BLACK);
+    init_pair(YELLOW_PAIR, COLOR_YELLOW, COLOR_BLACK);
+    init_pair(CYAN_PAIR, COLOR_CYAN, COLOR_BLACK);
+    return 0;
+}
+
+int render_board(gcb_t *gcb)
+{
+    mvprintw(11, 11, "+");
+    mvprintw(6, 21, "+");
+    mvprintw(1, 0, "   0 1 2 3 4 5 6 7 8 9 a b c d  │\n");
+    for (int y = 0; y < N_ROW; ++y) {
+        printw(" %x ", y);
+        for (int x = 0; x < N_COL; ++x) {
+            if (gcb->map[y][x] != -1) {
+                printw("%s ", (gcb->map[y][x])? "□" : "■"); 
+            } else {
+                printw("%s ", (((y == 4 && x == 9) || (y == 9 && x == 4)))? "+" : "·"); 
+            }
+        }
+        printw(" │\n");
+    }
+    return 0;
+}
+    
+int render_board_preview(rcb_t *rcb)
+{
+    gcb_t *gcb = rcb->gcb;
+    tile_t *tile = rcb->chosen;
+    int has_conflict = 0;
+    for (int i = 0; i < TILE[tile->shape].blk_cnt; ++i) {
+        int blk_x = tile->pos.x + tile->blks[i].x;
+        int blk_y = tile->pos.y + tile->blks[i].y;
+        if (gcb->map[blk_y][blk_x] != -1) {  // has block on such pos
+            has_conflict = 1; 
+        }
+    }
+    
+    start_color();
+    init_pair(RED_PAIR, COLOR_RED, COLOR_BLACK);
+    init_pair(GREEN_PAIR, COLOR_GREEN, COLOR_BLACK);
+    if (has_conflict) {
+        attron(COLOR_PAIR(RED_PAIR));
+    } else {
+        attron(COLOR_PAIR(GREEN_PAIR));
+    }
+    for (int i = 0; i < TILE[tile->shape].blk_cnt; ++i) {
+        int blk_x = tile->pos.x + tile->blks[i].x;
+        int blk_y = tile->pos.y + tile->blks[i].y;
+        mvprintw(blk_y + 2, blk_x * 2 + 3, "■");
+    }
+    if (has_conflict) {
+        attroff(COLOR_PAIR(RED_PAIR));
+    } else {
+        attroff(COLOR_PAIR(GREEN_PAIR));
+    }
+    return 0;
+}
+
+int recover_board_preview(rcb_t *rcb)
+{
+    gcb_t *gcb = rcb->gcb;
+    tile_t *tile = rcb->chosen;
+    for (int i = 0; i < TILE[tile->shape].blk_cnt; ++i) {
+        int blk_x = tile->pos.x + tile->blks[i].x;
+        int blk_y = tile->pos.y + tile->blks[i].y;
+        if (gcb->map[blk_y][blk_x] != -1) {
+            mvprintw(blk_y + 2, blk_x * 2 + 3, "%s ",
+                    (gcb->map[blk_y][blk_x])? "□" : "■"); 
```

Comment:
Let's align to the next char of the open parenthesis consistently.
```suggestion
            mvprintw(blk_y + 2, blk_x * 2 + 3, "%s ",
                     (gcb->map[blk_y][blk_x])? "□" : "■"); 
```